packages/jaeger-ui/src/model/link-patterns.js

```diff
@@ -0,0 +1,215 @@
+// @flow
+
+// Copyright (c) 2017 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import _uniq from 'lodash/uniq';
+import { getConfigValue } from '../utils/config/get-config';
+import type { Span, Trace } from '../types';
+
+const parameterRegExp = /\$\{([^{}]*)\}/g;
+
+type ProcessedTemplate = {
+  parameters: string[],
+  template: (data: { [parameter: string]: any }) => string,
+};
+
+function getParamNames(str) {
+  const names = [];
+  str.replace(parameterRegExp, (match, name) => {
+    names.push(name);
+    return match;
+  });
+  return _uniq(names);
+}
+
+function stringSupplant(str, encodeFn: any => string, map) {
+  return str.replace(parameterRegExp, (_, name) => encodeFn(map[name]));
+}
+
+export function processTemplate(template: any, encodeFn: any => string): ProcessedTemplate {
+  if (typeof template !== 'string') {
+    /*
+
+    // kept on ice until #123 is implemented:
+    if (template && Array.isArray(template.parameters) && (typeof template.template === 'function')) {
+      return template;
+    }
+
+    */
+    throw new Error('Invalid template');
+  }
+  return {
+    parameters: getParamNames(template),
+    template: stringSupplant.bind(null, template, encodeFn),
+  };
+}
+
+export function createTestFunction(entry: any) {
+  if (typeof entry === 'string') {
+    return (arg: any) => arg === entry;
+  }
+  if (Array.isArray(entry)) {
+    return (arg: any) => entry.indexOf(arg) > -1;
+  }
+  /*
+
+  // kept on ice until #123 is implemented:
+  if (entry instanceof RegExp) {
+    return (arg: any) => entry.test(arg);
+  }
+  if (typeof entry === 'function') {
+    return entry;
+  }
+
+  */
+  if (entry == null) {
+    return () => true;
+  }
+  throw new Error(`Invalid value: ${entry}`);
+}
+
+const identity = a => a;
+
+type ProcessedLinkPattern = {
```

Comment:
Can you please move type definitions to the top of the file?

Reply:
Ok!